File: examples/calc/ir-spec/calc_core.ml
type expr'0 = expr

and expr =
  | Expr'Number of int
  | Expr'Add of expr'0 * expr'0
  | Expr'Subtr of expr'0 * expr'0
  | Expr'Mult of expr'0 * expr'0
  | Expr'Divd of expr'0 * expr'0
  | Expr'Var of Ffi.atom'0

let rec expr_number'1 : int -> expr'0 = function v_N -> Expr'Number v_N

let rec expr_add'2 : expr'0 * expr'0 -> expr'0 = function
  | v_E1, v_E2 -> Expr'Add (v_E1, v_E2)

let rec expr_subtr'2 : expr'0 * expr'0 -> expr'0 = function
  | v_E1, v_E2 -> Expr'Subtr (v_E1, v_E2)

let rec expr_mult'2 : expr'0 * expr'0 -> expr'0 = function
  | v_E1, v_E2 -> Expr'Mult (v_E1, v_E2)

let rec expr_divd'2 : expr'0 * expr'0 -> expr'0 = function
  | v_E1, v_E2 -> Expr'Divd (v_E1, v_E2)

let rec expr_var'1 : Ffi.atom'0 -> expr'0 = function v_A -> Expr'Var v_A

File: examples/calc/ir-spec/calc_core.erl
-file("calc/src/calc_core.erlt", 1).


-module(calc_core).


-export_type([expr/0]).


-export([expr_number/1,
         expr_add/2,
         expr_subtr/2,
         expr_mult/2,
         expr_divd/2,
         expr_var/1]).


-type expr() ::
          {969696, calc_core, expr, number, integer()} |
          {969696, calc_core, expr, add, expr(), expr()} |
          {969696, calc_core, expr, subtr, expr(), expr()} |
          {969696, calc_core, expr, mult, expr(), expr()} |
          {969696, calc_core, expr, divd, expr(), expr()} |
          {969696, calc_core, expr, var, atom()}.


-spec expr_number(integer()) -> expr().


expr_number(N) ->
    {969696, calc_core, expr, number, N}.


-spec expr_add(expr(), expr()) -> expr().


expr_add(E1, E2) ->
    {969696, calc_core, expr, add, E1, E2}.


-spec expr_subtr(expr(), expr()) -> expr().


expr_subtr(E1, E2) ->
    {969696, calc_core, expr, subtr, E1, E2}.


-spec expr_mult(expr(), expr()) -> expr().


expr_mult(E1, E2) ->
    {969696, calc_core, expr, mult, E1, E2}.


-spec expr_divd(expr(), expr()) -> expr().


expr_divd(E1, E2) ->
    {969696, calc_core, expr, divd, E1, E2}.


-spec expr_var(atom()) -> expr().


expr_var(A) ->
    {969696, calc_core, expr, var, A}.





